### PR TITLE
BUGFIX: SD-354 tooltip responsive

### PR DESCRIPTION
**Pull request by lanhq**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2421
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2211

# Related Jira tasks

- SD-354: https://jira.intgdc.com/browse/SD-354

### CODE REVIEW

**Comment by lanhq** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1864/ : FAILURE in 1m 27s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5919/ : SUCCESS in 10s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4632/ : SUCCESS in 8m 38s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5918/ : SUCCESS in 10s

**Comment by lanhq** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- gooddata-react-components-pull-request-dispatcher : SKIPPED
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5920/ : Tests on opensource repositories should be triggered by users in `gooddata` organization in 0s

**Comment by lanhq** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1865/ : FAILURE in 25s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5921/ : SUCCESS in 8s

**Comment by lanhq** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2478/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1867/ : SUCCESS in 14m 11s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5924/ : SUCCESS in 8s

**Comment by lanhq** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1147.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1868/ : FAILURE in 22m 11s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5925/ : SUCCESS in 8s

**Comment by lanhq** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1147.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build in pipeline check-extended aborted.

File: package.json [hunk 1]
```diff
@@ -121,6 +121,7 @@
     "http-status-codes": "^1.3.0",
     "immer": "^1.3.1",
     "invariant": "^2.2.2",
+    "jest-canvas-mock": "^2.1.0",
```
Comment (ptquang86) [inline package.json]: this should be added into `devDependencies`, right?

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 2]
```diff
@@ -732,44 +737,66 @@ export function getSeries(
 
 export const customEscape = (str: string) => str && escape(unescape(str));
 
-const renderTooltipHTML = (textData: string[][], chartType: string, chartWidth: number): string => {
-    let tableStyle = "";
-    let columnTitleStyle = "";
-    let columnValueStyle = "";
+function isTextTooLong(textData: string[][], titleMaxWidth: number, valueMaxWidth: number) {
+    // calculate title and value length in pixels
+    const fontFamily = `"Avenir", "Helvetica Neue", arial, sans-serif`;
```
Comment (ptquang86) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: why don't we use single quote?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4634/ : SUCCESS in 13m 12s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5927/ : SUCCESS in 6s

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 3]
```diff
@@ -732,44 +737,66 @@ export function getSeries(
...
+    const valueStyle = ` style="max-width: ${valueMaxWidth}px; padding-left: ${valuePaddingLeft}px;"`;
+
+    return {
+        htmlContent: `${textData
+            .map((item, index) => {
+                const lastClass = index === textData.length - 1 ? " last" : "";
+                return `<div class="gd-viz-tooltip-item${lastClass}">
+                <span class="gd-viz-tooltip-title"${titleStyle}>${item[0]}</span>
+                <span class="gd-viz-tooltip-value"${valueStyle}>${item[1]}</span>
+            </div>`;
```
Comment (ptquang86) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: Please fix the indents

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4635/ : SUCCESS in 5m 16s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5928/ : SUCCESS in 8s

File: src/components/visualizations/chart/chartOptionsBuilder.ts (hunk 3 above)
Comment (lanhq) [reply src/components/visualizations/chart/chartOptionsBuilder.ts]: updated

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4641/ : SUCCESS in 4m 46s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5934/ : SUCCESS in 8s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2480/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1871/ : SUCCESS in 14m 18s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5935/ : SUCCESS in 9s

**Comment by lanhq** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1147.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1872/ : FAILURE in 25m 21s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5936/ : SUCCESS in 12s

**Comment by lanhq** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1147.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4642/ : SUCCESS in 8m 02s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5937/ : SUCCESS in 14s

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1873/ : SUCCESS in 17m 47s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5938/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4647/ : FAILURE in 12m 34s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5945/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4648/ : SUCCESS in 4m 23s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5946/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4649/ : SUCCESS in 6m 21s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5947/ : SUCCESS in 9s

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 4]
```diff
@@ -732,44 +731,35 @@ export function getSeries(
...
-                <td class="gd-viz-tooltip-table-cell value gd-viz-tooltip-table-value"${columnValueStyle}>${
-                    line[1]
-                }</td>
-            </tr>`,
-        )
-        .join("\n")}</table>`;
+const renderTooltipHTML = (textData: string[][], maxTooltipContentWidth: number): string => {
+    const maxItemWidth = maxTooltipContentWidth - TOOLTIP_PADDING * 2;
+    const titleMaxWidth = maxItemWidth;
+    const threeDotWidth = 16;
```
Comment (imnutz) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: ```threeDotsWidth```

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4650/ : SUCCESS in 7m 41s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5948/ : SUCCESS in 10s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4651/ : SUCCESS in 8m 09s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5949/ : SUCCESS in 12s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4653/ : SUCCESS in 6m 36s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5951/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4662/ : SUCCESS in 8m 12s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5963/ : SUCCESS in 6s

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 5]
```diff
@@ -103,6 +96,9 @@ import { NORMAL_STACK, PERCENT_STACK } from "./highcharts/getOptionalStackingCon
 
 import { getCategoriesForTwoAttributes } from "./chartOptions/extendedStackingChartOptions";
 import { setMeasuresToSecondaryAxis } from "../../../helpers/dualAxis";
+import { isCssSupportMultiLineTruncation } from "../../../helpers/domUtils";
```
Comment (ptquang86) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: how about `isCssMultiLineTruncationSupported`?

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 6]
```diff
@@ -733,44 +733,40 @@ export function getSeries(
...
-        )
-        .join("\n")}</table>`;
+    const titleStyle = ` style="max-width: ${titleMaxWidth}px;"`;
+    const valueStyle = ` style="max-width: ${valueMaxWidth}px;"`;
+
+    return textData
+        .map((item, index) => {
+            const lastClass = index === textData.length - 1 ? "last" : "";
+            // the third span is hidden, that help to have tooltip work with max-width
+            return `<div class="gd-viz-tooltip-item ${supportMultilineTrucationClass} ${lastClass}">
```
Comment (ptquang86) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: should we re-use `classnames` module?

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4664/ : SUCCESS in 6m 11s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5965/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4665/ : SUCCESS in 5m 35s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5966/ : SUCCESS in 11s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4667/ : SUCCESS in 9m 45s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5968/ : SUCCESS in 46s

File: src/components/visualizations/chart/chartOptionsBuilder.ts [hunk 7]
```diff
@@ -733,44 +734,38 @@ export function getSeries(
...
-
-        tableStyle = ` style="max-width: ${tableMaxWidth}px;-webkit-border-horizontal-spacing: 0;"`;
-        columnTitleStyle = ` style="max-width: ${columnMaxWidth}px;"`;
-        columnValueStyle = ` style="max-width: ${columnMaxWidth - TOOLTIP_VALUE_PADDING}px;"`;
-    }
+const renderTooltipHTML = (textData: string[][], maxTooltipContentWidth: number): string => {
+    const maxItemWidth = maxTooltipContentWidth - TOOLTIP_PADDING * 2;
+    const titleMaxWidth = maxItemWidth;
+    const multiLineTruncationSupported = isCssMultiLineTruncationSupported();
+    const threeDotsWidth = 16;
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/chartOptionsBuilder.ts]: does it work with zoomed in browser?

File: src/components/visualizations/chart/highcharts/customConfiguration.ts [hunk 8]
```diff
@@ -69,7 +62,7 @@ const ALIGN_CENTER = "center";
 
 const TOOLTIP_ARROW_OFFSET = 23;
 const TOOLTIP_FULLSCREEN_THRESHOLD = 480;
-const TOOLTIP_MAX_WIDTH = 366;
+const TOOLTIP_MAX_WIDTH = 320;
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/chart/highcharts/customConfiguration.ts]: Who provided this new value?

File: src/components/visualizations/styles/tooltip.scss [hunk 9]
```diff
@@ -86,72 +86,90 @@
...
-    white-space: nowrap;
-    text-overflow: ellipsis;
-    vertical-align: top;
+.multiline-supported {
+    .gd-viz-tooltip-value-wraper {
+        padding-right: 0;
+    }
+
+    .gd-viz-tooltip-value {
+        display: -webkit-box;
```
Comment (ivan-nejezchleb) [inline src/components/visualizations/styles/tooltip.scss]: is it possible to keep 'display', '-webkit-line-clamp' and '-webkit-box-orient' props next to each other or the best in separate class to explicitly say that they need to be set together to allow clamping to work?

File: src/components/visualizations/chart/chartOptionsBuilder.ts (hunk 7 above)
Comment (lanhq) [reply src/components/visualizations/chart/chartOptionsBuilder.ts]: i checked & it still work well

File: src/components/visualizations/chart/highcharts/customConfiguration.ts (hunk 8 above)
Comment (lanhq) [reply src/components/visualizations/chart/highcharts/customConfiguration.ts]: it is current max tooltip content width (see [old css](https://github.com/gooddata/gooddata-react-components/pull/1147/files#diff-84b5bf43cbbf7caee4c0bdabb659f0efL119)). i just adjust the border width and padding as [mockup](https://gooddata.invisionapp.com/share/5MSHNVYCSUQ#/screens/368579121)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4672/ : SUCCESS in 6m 58s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5976/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4673/ : SUCCESS in 4m 12s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5977/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4674/ : SUCCESS in 5m 11s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5978/ : SUCCESS in 23s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4675/ : SUCCESS in 4m 05s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5979/ : SUCCESS in 16s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4676/ : SUCCESS in 9m 03s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5980/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2492/ : SUCCESS in 11m 16s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4679/ : SUCCESS in 9m 40s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1184/ : SUCCESS in 25s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1520/ : SUCCESS in 15m 41s

**Pull request merged**